backtesting/backtesting.py

```diff
@@ -1221,101 +1236,172 @@ def maximize(stats: pd.Series, _key=maximize):
+            """
+            Optimize strategy parameters to an optimal combination using
+            scikit-optimize. Returns result `pd.Series` of
+            the best run.
+
+            `maximize` is a string key from the
+            `backtesting.backtesting.Backtest.run`-returned results series,
+            or a function that accepts this series object and returns a number;
+            the higher the better. By default, the method maximizes
+            Van Tharp's [System Quality Number](https://google.com/search?q=System+Quality+Number).
+
+            Additional keyword arguments represent strategy arguments with
+            list-like collections of possible values. For example, the following
+            code finds and returns the "best" of the 7 admissible (of the
+            9 possible) parameter combinations:
+
+                backtest.optimize(sma1=[5, 10, 15], sma2=[10, 20, 40],
+                                constraint=lambda p: p.sma1 < p.sma2)
+
+            .. TODO::
+                Improve multiprocessing with joblib.parallel .
+            """
+            dimensions = []
+            for key, value in kwargs.items():
+                args_array = np.array(value)
+
+                if args_array.dtype.kind == 'i':
+                    dimensions.append(Integer(name=key, low=min(value), high=max(value)))
+                elif args_array.dtype.kind == 'f':
+                    dimensions.append(Real(name=key, low=min(value), high=max(value)))
+                else:
+                    dimensions.append(Categorical(args_array.tolist(), name=key))
+
+            # @ use_named_args(dimensions=dimensions)
+            # def convert(n1, n2):
+            #     res = self.run(**dict(zip(['n1', 'n2'], [n1, n2])))
+            #     return -res[maximize_key]
+            @ use_named_args(dimensions=dimensions)
+            def convert(**params):
+                index = []
+                for item in params:
+                    index.append("'" + item + "'")
+                res = self.run(**dict(zip(index, params)))
+                return -res[maximize_key]
+
+            res = forest_minimize(func=convert,
+                                  dimensions=dimensions,
+                                  n_calls=max_tries, base_estimator="ET",
+                                  random_state=4)
+
+            self.run(**dict(zip(['n1', 'n2'], res.x)))
+            return_result = pd.Series([res.fun, res.x, res, self._results], index=[
+                "best_fitness", "best_parameters", "full_results", "stats_best"])
+
+            if return_heatmap:
+                return return_result["stats_best"], return_result["full_results"]
+            return return_result["stats_best"]
+
+        def switch_method(x):
+            return {
+                'auto': _auto(),
+                'full': 0,
+                'random': 0,
+                'skopt': _optimize_skopt(),
```

Comment:
Nice try. :smile: I actually meant `'auto'` as a default, to be a synonym for:
```py
'full' if len(param_combos) <= max_tries else 'random' if skopt is None else 'skopt'
```
The question is, why would we construct `param_combos` that can potentially explode and that we don't have intention of using. But if one is using `method='skopt'`, it suffices they provide just the min and max of ordinal values, so the parameter space might not be that large after all.

Reply:
Okay, I have renamed the `auto` to `grid` which is the default method. As you mentioned, we don't need to build param_combos if it is not used in skopt (or likely any non-grid method) .

Reply:
'auto' is somewhat an [industry standard](https://scikit-learn.org/stable/search.html?q=%27auto%27), indicating that the value to be used depends on some aspect of the passed data.

Reply:
Finally opted for no `'auto'`. Switching methods is simple enough.